Support RTree index by boost

xiaocai2333 opened the pull request.

Yinwei-Yu committed:
Add geometry support.

Support geometry in milvus.Now can insert,query and search with gis functions.
This commit includes file changing with related to 3rdparty dependencies,cpp core and go.
Here is also an example file which shows how to use the geo type and gis functions.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

xiaocai2333 committed:
Debug geo index

Signed-off-by: Cai Zhang <cai.zhang@zilliz.com>

xiaocai2333 commented on internal/core/src/index/Meta.h:
```diff
@@ -93,4 +94,9 @@ constexpr const char* DISK_ANN_PREPARE_USE_BFS_CACHE = "use_bfs_cache";
 constexpr const char* DISK_ANN_QUERY_LIST = "search_list";
 constexpr const char* DISK_ANN_QUERY_BEAMWIDTH = "beamwidth";
+
+constexpr const char* R_TREE_VARIANT_KEY = "rv";
```
need remove this

xiaocai2333 commented on internal/core/src/index/Utils.cpp:
```diff
@@ -251,6 +251,56 @@ GetIndexMetaFromConfig(const Config& config) {
     return index_meta;
 }
 
+double
```
ditto

Yinwei-Yu commented on client/index/rtree.go:
```diff
@@ -0,0 +1,148 @@
+	RTreeLeafCapacityKey  = "leafCapacity"
+	RTreeDimKey           = "dim"
+	RTreeRVKey            = "rv"
+)
```
Since we changed to boost/geometry, so the params here for libspatialindex is useless ,and so is rtree_test.go

xiaocai2333 commented on internal/util/indexparamcheck/constraints.go:
```diff
@@ -49,6 +49,32 @@ const (
 
 	MaxBitmapCardinalityLimit = 1000
+
+	// RTree Index Param
```
ditto

xiaocai2333 reviewed (commented):
I will do this.

Yinwei-Yu commented on configs/milvus.yaml:
```diff
@@ -787,7 +787,7 @@ msgChannel:
   # Milvus log level. Option: debug, info, warn, error, panic, and fatal. 
   # It is recommended to use debug level under test and development environments, and info level in production environment.
-  level: info
+  level: debug
```
use info

Yinwei-Yu commented on internal/core/src/index/Meta.h:
```diff
@@ -93,4 +94,9 @@ constexpr const char* DISK_ANN_PREPARE_USE_BFS_CACHE = "use_bfs_cache";
 constexpr const char* DISK_ANN_QUERY_LIST = "search_list";
 constexpr const char* DISK_ANN_QUERY_BEAMWIDTH = "beamwidth";
+
+constexpr const char* R_TREE_VARIANT_KEY = "rv";
+constexpr const char* FILL_FACTOR_KEY = "fillFactor";
+constexpr const char* INDEX_CAPACITY_KEY = "indexCapacity";
+constexpr const char* LEAF_CAPACITY_KEY = "leafCapacity";
```
useless for now

Yinwei-Yu commented on internal/core/src/index/RTreeIndexWrapper.h:
```diff
@@ -0,0 +1,143 @@
+ * @brief Wrapper class for libspatialindex R-Tree functionality
+ * 
+ * This class provides a simplified interface to libspatialindex library,
+ * handling the creation, management, and querying of R-Tree spatial indexes
```
Change the comment from libspatialindex to boost

Yinwei-Yu commented on internal/core/src/index/RTreeIndexWrapper.h:
```diff
@@ -0,0 +1,143 @@
+
+ private:
+    // Boost.Geometry types and in-memory structures
+    using Point = bg::model::point<double, 2, bg::cs::cartesian>;
```
change this to `using Point = bg::model::point<double, 2, bg::cs::geographic<bg::degree>>;`

Yinwei-Yu commented on internal/core/src/index/RTreeIndex.cpp:
```diff
@@ -0,0 +1,531 @@
+            base_path = p.substr(0, p.size() - 4);
+            break;
+        }
+    }
```
This is used for libspatialindex, so delete it

Yinwei-Yu commented on internal/core/src/index/Utils.h:
```diff
@@ -163,6 +163,15 @@ GetFieldDataMetaFromConfig(const Config& config);
 storage::IndexMeta
 GetIndexMetaFromConfig(const Config& config);
 
+double
+GetFillFactorFromConfig(const Config& config);
+
+uint32_t
+GetIndexCapacityFromConfig(const Config& config);
+
+uint32_t
+GetLeafCapacityFromConfig(const Config& config);
+
```
delete this

Yinwei-Yu commented on internal/core/unittest/test_rtree_index.cpp:
All the test related with libspatialindex should be changed

Yinwei-Yu commented on internal/core/conanfile.py:
```diff
@@ -54,7 +54,8 @@ class MilvusConan(ConanFile):
         "geos/3.12.0#b76c27884c1fa4ee8c9e486337b7dc4e",
-        "gdal/3.5.3#61a42c933d3440a449cac89fd0866621"
+        "gdal/3.5.3#61a42c933d3440a449cac89fd0866621",
+        "libspatialindex/2.1.0#866b4d23930c42221f0f28547ed2b3d5"
```
delete this

Yinwei-Yu commented on internal/util/indexparamcheck/rtree_checker.go:
Change this file to align with boost

Yinwei-Yu commented on internal/core/conanfile.py:
```diff
@@ -6,7 +6,7 @@ class MilvusConan(ConanFile):
     requires = (
         "rocksdb/6.29.5@milvus/dev#b1842a53ddff60240c5282a3da498ba1",
-        "boost/1.82.0#744a17160ebb5838e9115eab4d6d0c06",
+        "boost/1.83.0@",
```
Why doesn't here need a SHA?

xiaocai2333 replied:
<img width="1050" height="665" alt="image" src="https://github.com/user-attachments/assets/3b8edf55-29e2-452b-9ad1-144bcb6c53ea" />
there is no SHA with v1.83.0 in default-conan-local.

Yinwei-Yu reviewed (commented):
Please check

Yinwei-Yu committed:
support geometry and brute search (#5)

* Add geometry support.

Support geometry in milvus.Now can insert,query and search with gis functions.
This commit includes file changing with related to 3rdparty dependencies,cpp core and go.
Here is also an example file which shows how to use the geo type and gis functions.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>
Co-authored-by: cai.zhang <cai.zhang@zilliz.com>
